### PR TITLE
[CUBRIDQA-40] enhance continue mode for shell tool

### PR DESCRIPTION
**Pull request by supler**
@fanzaiqiang 

I add one method to judge if it needs continue testing, please review it. thanks

### CODE REVIEW

File: CTP/shell/src/com/navercorp/cubridqa/shell/main/TestFactory.java [hunk 1]
```diff
@@ -83,7 +91,7 @@ private void addSkippedTestCases(ArrayList<String> list, String skippedType) {
 
 	public void execute() throws Exception {
```
Comment (fanzaiqiang) [inline CTP/shell/src/com/navercorp/cubridqa/shell/main/TestFactory.java]: Maybe we do not need isNeedContinue() method.
Just consider an existing method as below:
if (context.isContinueMode) {
    Dispatch.init(context);
    if(Dispatch.getInstance().getTotalTbdSize() == 0 ) {
        System.out.println("Not found any test cases to execute with CONTINUE mode");
        return;
    }
}

**Comment by supler** [comment]
@fanzaiqiang 

implement failures backup in java and use the existing method to implement continue mode, please review it again. thanks

File: CTP/shell/src/com/navercorp/cubridqa/shell/common/CommonUtils.java [hunk 2]
```diff
@@ -280,6 +283,25 @@ public static String toFixedLength(String str, int len, char fillChar) {
 		return result.substring(result.length() - len);
 	}
 	
+	public static void generateFailBackupPackage(Context context){
+		if(context.isWindows()) return;
```
Comment (fanzaiqiang) [inline CTP/shell/src/com/navercorp/cubridqa/shell/common/CommonUtils.java]: as is:
if(context.isWindows()) return;

The context.isWindows() always means the test node is in windows platform. It should not be used to controller platform.

File: CTP/shell/src/com/navercorp/cubridqa/shell/common/CommonUtils.java [hunk 3]
```diff
@@ -280,6 +283,25 @@ public static String toFixedLength(String str, int len, char fillChar) {
...
+		int year = cal.get(Calendar.YEAR);  
+        int month = cal.get(Calendar.MONTH) + 1;
+        int day = cal.get(Calendar.DAY_OF_MONTH);  
+        int hour = cal.get(Calendar.HOUR);  
+        int minute = cal.get(Calendar.MINUTE);  
+        int second = cal.get(Calendar.SECOND);
+        String curTimestamp = year + "." + month + "." + day + "_" + hour + "." + minute + "." + second;
+        String build = context.getTestBuild();
+        String bit = context.getVersion();
+        Integer tastId = context.getTaskId();
```
Comment (fanzaiqiang) [inline CTP/shell/src/com/navercorp/cubridqa/shell/common/CommonUtils.java]: as is:
 Integer tastId = context.getTaskId();

There is typo in 'tastId'.

And it's better to apply the code format.

File: CTP/shell/src/com/navercorp/cubridqa/shell/common/CommonUtils.java [hunk 4]
```diff
@@ -280,6 +283,25 @@ public static String toFixedLength(String str, int len, char fillChar) {
...
+        
+        backupFileName = "shell_result_" + build + "_" + bit + "_" + tastId + "_" + curTimestamp + "tar.gz";
+        LocalInvoker.exec("tar zvcf " + Constants.DIR_WORKS_ROOT + "/" + backupFileName + " conf log run.log " , false, false);
```
Comment (fanzaiqiang) [inline CTP/shell/src/com/navercorp/cubridqa/shell/common/CommonUtils.java]: Please package to CTP_HOME/result/shell/ directory.

File: CTP/shell/src/com/navercorp/cubridqa/shell/common/Constants.java [hunk 5]
```diff
@@ -37,6 +37,7 @@
 	
 	public static final String DIR_CONF = "./conf";
```
Comment (fanzaiqiang) [inline CTP/shell/src/com/navercorp/cubridqa/shell/common/Constants.java]: Maybe we should refactor DIR_CONF, DIR_LOG_ROOT, DIR_WORKS_ROOT folders.

File: CTP/shell/src/com/navercorp/cubridqa/shell/main/TestFactory.java [hunk 6]
```diff
@@ -82,24 +90,31 @@ private void addSkippedTestCases(ArrayList<String> list, String skippedType) {
...
-			System.out.println("NO TEST CASE TO TEST");			
-			return;
+			System.out.println("============= SVN UPDATE ==================");
+			concurrentSVNUpdate(stableEnvList);
+			System.out.println("DONE");
+			
+			System.out.println("============= FETCH TEST CASES ==================");
+			Dispatch.init(context);
+			if (Dispatch.getInstance().getTotalTbdSize() == 0) {
+				System.out.println("NO TEST CASE TO TEST!");
```
Comment (fanzaiqiang) [inline CTP/shell/src/com/navercorp/cubridqa/shell/main/TestFactory.java]: System.out.println("Not found any test cases to execute with CONTINUE mode!");
System.out.println("NO TEST CASE TO TEST!");

It's better to have consistent output information.

**Comment by supler** [comment]
@fanzaiqiang 

I refine the codes according to the comments, please review it again. thanks

File: CTP/.classpath [hunk 7]
```diff
@@ -46,5 +46,6 @@
 	<classpathentry kind="lib" path="common/sched/lib/geronimo-j2ee-management_1.1_spec-1.0.1.jar"/>
 	<classpathentry kind="lib" path="common/sched/lib/hawtbuf-1.9.jar"/>
 	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER"/>
+	<classpathentry kind="lib" path="common/lib/cubridqa-common.jar"/>
```
Comment (fanzaiqiang) [inline CTP/.classpath]: Please remove:
<classpathentry kind="lib" path="common/lib/cubridqa-common.jar"/>

File: CTP/common/src/com/navercorp/cubridqa/common/Constants.java [hunk 8]
```diff
@@ -32,6 +32,9 @@
 	public final static String LINE_SEPARATOR = System.getProperty("line.separator");
 	
 	public final static Properties COMMON_DAILYQA_CONF;
+	
+	public static final String TOOL_HOME_NAME="CTP_HOME";
```
Comment (fanzaiqiang) [inline CTP/common/src/com/navercorp/cubridqa/common/Constants.java]: Rename TOOL_HOME_NAME to ENV_CTP_HOME_KEY

File: CTP/shell/src/com/navercorp/cubridqa/shell/result/FeedbackFile.java [hunk 9]
```diff
@@ -47,7 +47,7 @@
 	Context context;
 	
 	public FeedbackFile(Context context){
-		logName = CommonUtils.concatFile(Constants.DIR_LOG_ROOT, "feedback.log");
+		logName = CommonUtils.concatFile(context.getToolHome() + "/" + Constants.CURRENT_LOG_DIR, "feedback.log");
```
Comment (fanzaiqiang) [inline CTP/shell/src/com/navercorp/cubridqa/shell/result/FeedbackFile.java]: Context.getCurrentLogDir() {
        return getToolHome() + "/" + Constants.CURRENT_LOG_DIR
}

File: CTP/shell/src/com/navercorp/cubridqa/shell/result/FeedbackDB.java [hunk 10]
```diff
@@ -111,8 +111,9 @@ public void onTaskStartEvent(String buildFilename) {
 			close(conn);
 		}
 
-		Log log = new Log(CommonUtils.concatFile(Constants.DIR_CONF, "current_task_id"), false, false);
```
Comment (fanzaiqiang) [inline CTP/shell/src/com/navercorp/cubridqa/shell/result/FeedbackDB.java]: Change to use context.getCurrentLogDir() to get the directory.

File: CTP/shell/src/com/navercorp/cubridqa/shell/main/Test.java [hunk 11]
```diff
@@ -72,8 +72,8 @@ public Test(Context context, String currEnvId, boolean laterJoined) throws JSchE
 		this.context = context;
 		this.currEnvId = currEnvId;
 		this.startTime = 0;
-		this.dispatchLog = new Log(CommonUtils.concatFile(Constants.DIR_CONF, "dispatch_tc_FIN_" + currEnvId + ".txt"), false, laterJoined ? true : context.isContinueMode());
-		this.workerLog = new Log(CommonUtils.concatFile(Constants.DIR_LOG_ROOT, "test_" + currEnvId + ".log"), false, true);
+		this.dispatchLog = new Log(CommonUtils.concatFile(context.getToolHome() + "/" + Constants.CURRENT_LOG_DIR, "dispatch_tc_FIN_" + currEnvId + ".txt"), false, laterJoined ? true : context.isContinueMode());
```
Comment (fanzaiqiang) [inline CTP/shell/src/com/navercorp/cubridqa/shell/main/Test.java]: Change to use context.getCurrentLogDir() to get the directory.

File: CTP/shell/src/com/navercorp/cubridqa/shell/main/Main.java [hunk 12]
```diff
@@ -78,37 +79,45 @@ public static void exec(String configFilename) throws Exception {
...
 		}
 
 		Properties props = context.getProperties();
 		Set set = props.keySet();
-		Log contextSnapshot = new Log(CommonUtils.concatFile(Constants.DIR_CONF, "main_snapshot.properties"), true, false);
-		for(Object key: set) {
-			contextSnapshot.println(key + "=" + props.getProperty((String)key) );
+		Log contextSnapshot = new Log(
+				CommonUtils
+						.concatFile(context.getToolHome() + "/"
```
Comment (fanzaiqiang) [inline CTP/shell/src/com/navercorp/cubridqa/shell/main/Main.java]: Change to use context.getCurrentLogDir() to get the directory.

File: CTP/shell/src/com/navercorp/cubridqa/shell/common/CommonUtils.java [hunk 13]
```diff
@@ -120,12 +124,20 @@ public static String concatFile(String p1, String p2) {
...
+	public static String getFileNameForDispatchFin(String envId, Context context){
+		return CommonUtils.concatFile(context.getToolHome() + "/" + Constants.CURRENT_LOG_DIR,"dispatch_tc_FIN_" + envId + ".txt");
+	}
+	
+	public static String getEnvInMemory(String key){
+		return com.navercorp.cubridqa.common.CommonUtils.getEnvInFile(key);
+	}
+	
+	public static boolean isWindowsPlatformForController(){
+		return com.navercorp.cubridqa.common.CommonUtils.isCygwinPlatform();
```
Comment (fanzaiqiang) [inline CTP/shell/src/com/navercorp/cubridqa/shell/common/CommonUtils.java]: return com.navercorp.cubridqa.common.CommonUtils.isWindowsPlatform();

File: CTP/shell/src/com/navercorp/cubridqa/shell/common/CommonUtils.java [hunk 14]
```diff
@@ -120,12 +124,20 @@ public static String concatFile(String p1, String p2) {
...
+		return CommonUtils.concatFile(context.getToolHome() + "/" + Constants.CURRENT_LOG_DIR, "dispatch_tc_ALL.txt");
 	}
 
-	public static String getFileNameForDispatchFin(String envId){
-		return CommonUtils.concatFile(Constants.DIR_CONF,"dispatch_tc_FIN_" + envId + ".txt");
+	public static String getFileNameForDispatchFin(String envId, Context context){
+		return CommonUtils.concatFile(context.getToolHome() + "/" + Constants.CURRENT_LOG_DIR,"dispatch_tc_FIN_" + envId + ".txt");
+	}
+	
+	public static String getEnvInMemory(String key){
```
Comment (fanzaiqiang) [inline CTP/shell/src/com/navercorp/cubridqa/shell/common/CommonUtils.java]: It's better remove getEnvInMemory().

Just use com.navercorp.cubridqa.common.CommonUtils.getEnvInFile(key) as required.

File: CTP/shell/src/com/navercorp/cubridqa/shell/common/CommonUtils.java [hunk 15]
```diff
@@ -120,12 +124,20 @@ public static String concatFile(String p1, String p2) {
 		return p.replace('/', File.separatorChar);
 	}
 	
-	public static String getFileNameForDispatchAll(){
-		return CommonUtils.concatFile(Constants.DIR_CONF, "dispatch_tc_ALL.txt");
+	public static String getFileNameForDispatchAll(Context context){
```
Comment (fanzaiqiang) [inline CTP/shell/src/com/navercorp/cubridqa/shell/common/CommonUtils.java]: Move below two methods to Dispatch.java

public static String getFileNameForDispatchAll(Context context)
public static String getFileNameForDispatchFin(String envId, Context context){

File: CTP/shell/src/com/navercorp/cubridqa/shell/common/CommonUtils.java [hunk 16]
```diff
@@ -280,6 +292,30 @@ public static String toFixedLength(String str, int len, char fillChar) {
...
+		String curTimestamp = year + "." + month + "." + day + "_" + hour + "."
+				+ minute + "." + second;
+		String build = context.getTestBuild();
+		String bit = context.getVersion();
+		Integer taskId = context.getTaskId();
+
+		backupFileName = "shell_result_" + build + "_" + bit + "_" + taskId
+				+ "_" + curTimestamp + "tar.gz";
+		LocalInvoker.exec(
+				"cd $CTP_HOME; tar zvcf " + Constants.RUNTIME_LOG_DIR_HOME + "/"
```
Comment (fanzaiqiang) [inline CTP/shell/src/com/navercorp/cubridqa/shell/common/CommonUtils.java]: Change to use context.getLogRootDir();

File: CTP/shell/src/com/navercorp/cubridqa/shell/common/Constants.java [hunk 17]
```diff
@@ -35,14 +35,14 @@
...
-	public static final String DIR_CONF = "./conf";
-	public static final String DIR_LOG_ROOT = "./log";
-	
 	public static final String TC_RESULT_OK = "OK";
 	public static final String TC_RESULT_NOK = "NOK";
 	
 	public static final String RETRY_FLAG = "Retry_Count";
-
+	
+	public static final String RUNTIME_LOG_DIR_HOME = "result/shell";
```
Comment (fanzaiqiang) [inline CTP/shell/src/com/navercorp/cubridqa/shell/common/Constants.java]: May remove both RUNTIME_LOG_DIR_HOME and CURRENT_LOG_DIR.

Replace with fetching from Context.java

**Comment by fanzaiqiang** [comment]
@supler 

Let's improve as my suggestions above. Thanks a lot.

**Comment by supler** [comment]
@fanzaiqiang 

I refine it again, please review it. thanks

File: CTP/shell/src/com/navercorp/cubridqa/shell/deploy/Deploy.java [hunk 18]
```diff
@@ -55,7 +55,7 @@ public Deploy(Context context, String currEnvId, boolean laterJoined) throws Exc
 
 		this.cubridPackageUrl = context.getCubridPackageUrl();
 
-		this.log = new Log(CommonUtils.concatFile(Constants.DIR_LOG_ROOT, "test_" + currEnvId + ".log"), false, laterJoined ? true : context.isContinueMode());
+		this.log = new Log(CommonUtils.concatFile(Constants.CURRENT_LOG_DIR, "test_" + currEnvId + ".log"), false, laterJoined ? true : context.isContinueMode());
```
Comment (fanzaiqiang) [inline CTP/shell/src/com/navercorp/cubridqa/shell/deploy/Deploy.java]: Change to use context.getCurrentLogDir().

File: CTP/shell/src/com/navercorp/cubridqa/shell/dispatch/Dispatch.java [hunk 19]
```diff
@@ -318,12 +323,20 @@ private static String getAllTestCaseScripts(String dir) {
...
+	private String getFileNameForDispatchFin(String envId){
+		return CommonUtils.concatFile(context.getCurrentLogDir(),"dispatch_tc_FIN_" + envId + ".txt");
+	}
 
 	private void clean() throws IOException {
 		File allFile;
 		File finishedFile;
 
-		File[] subList = new File(Constants.DIR_CONF).listFiles(new FilenameFilter() {
+		File[] subList = new File(Constants.CURRENT_LOG_DIR).listFiles(new FilenameFilter() {
```
Comment (fanzaiqiang) [inline CTP/shell/src/com/navercorp/cubridqa/shell/dispatch/Dispatch.java]: Change to use context.getCurrentLogDir().

File: CTP/shell/src/com/navercorp/cubridqa/shell/main/TestMonitor.java [hunk 20]
```diff
@@ -51,15 +51,16 @@ public TestMonitor(Context context, Test test) throws Exception {
 		this.context = context;
 		this.test = test;
 		
-		this.log = new Log(CommonUtils.concatFile(Constants.DIR_LOG_ROOT, "monitor_" + test.getCurrentEnvId() + ".log"), false, context.isContinueMode);
+		this.log = new Log(CommonUtils.concatFile(context.getToolHome() + "/" + Constants.CURRENT_LOG_DIR, "monitor_" + test.getCurrentEnvId() + ".log"), false, context.isContinueMode);
```
Comment (fanzaiqiang) [inline CTP/shell/src/com/navercorp/cubridqa/shell/main/TestMonitor.java]: Change to use context.getCurrentLogDir()

File: CTP/shell/src/com/navercorp/cubridqa/shell/main/TestMonitor.java [hunk 21]
```diff
@@ -76,7 +77,7 @@ public TestMonitor(Context context, Test test) throws Exception {
 		}
 		
 		if (enableTracing && sshRelateds != null && sshRelateds.size() > 0) {
-			this.logRelated = new Log(CommonUtils.concatFile(Constants.DIR_LOG_ROOT, "monitor_" + test.getCurrentEnvId() + "_related.log"), false, context.isContinueMode);
+			this.logRelated = new Log(CommonUtils.concatFile(context.getToolHome() + "/" + Constants.CURRENT_LOG_DIR, "monitor_" + test.getCurrentEnvId() + "_related.log"), false, context.isContinueMode);
```
Comment (fanzaiqiang) [inline CTP/shell/src/com/navercorp/cubridqa/shell/main/TestMonitor.java]: Change to use context.getCurrentLogDir().

File: CTP/shell/src/com/navercorp/cubridqa/shell/result/FeedbackDB.java [hunk 22]
```diff
@@ -121,8 +122,9 @@ public void onTaskContinueEvent() {
 
 		String cont = null;
 		try {
-			cont = CommonUtils.getFileContent(CommonUtils.concatFile(Constants.DIR_CONF, "current_task_id"));
+			cont = CommonUtils.getFileContent(CommonUtils.concatFile(context.getToolHome() + "/" + Constants.CURRENT_LOG_DIR, "current_task_id"));
```
Comment (fanzaiqiang) [inline CTP/shell/src/com/navercorp/cubridqa/shell/result/FeedbackDB.java]: Change to use context.getCurrentLogDir().

**Comment by supler** [comment]
@fanzaiqiang 

I global checking for log directory and update them as getCurrentLogDir() to get

**Comment by fanzaiqiang** [comment]
@supler 

Please go ahead to merge it. Thanks a lot.

**Comment by supler** [comment]
this changes were merge from #66,so I close this request

**Pull request closed without merging**